revocation/revocation.go

```diff
@@ -56,26 +78,140 @@ func NewTimestamp(httpClient *http.Client) (Revocation, error) {
+// CRL if OCSP is not available. It returns an array of CertRevocationResults
+// that contain the results and any errors that are encountered during the
+// process.
+//
+// The certificate chain is expected to be in the order of leaf to root.
+//
+// This function tries OCSP and falls back to CRL when:
+// - OCSP is not supported by the certificate
+// - OCSP returns an unknown status
 //
-// TODO: add CRL support
-// https://github.com/notaryproject/notation-core-go/issues/125
+// When OCSP returns an unknown status, the function will try to check the
+// certificate status using CRL and return certificate result with an
+// result.OCSPFallbackError.
 func (r *revocation) Validate(certChain []*x509.Certificate, signingTime time.Time) ([]*result.CertRevocationResult, error) {
-	return ocsp.CheckStatus(ocsp.Options{
+	if len(certChain) == 0 {
+		return nil, result.InvalidChainError{Err: errors.New("chain does not contain any certificates")}
+	}
+
+	// Validate cert chain structure
+	// Since this is using authentic signing time, signing time may be zero.
+	// Thus, it is better to pass nil here than fail for a cert's NotBefore
+	// being after zero time
+	switch r.certChainPurpose {
+	case ocsp.PurposeCodeSigning:
+		if err := coreX509.ValidateCodeSigningCertChain(certChain, nil); err != nil {
+			return nil, result.InvalidChainError{Err: err}
+		}
+	case ocsp.PurposeTimestamping:
+		if err := coreX509.ValidateTimestampingCertChain(certChain); err != nil {
+			return nil, result.InvalidChainError{Err: err}
+		}
+	default:
+		return nil, result.InvalidChainError{Err: fmt.Errorf("unknown certificate chain purpose %v", r.certChainPurpose)}
+	}
+
+	ocspOpts := ocsp.Options{
 		CertChain:        certChain,
-		CertChainPurpose: r.certChainPurpose,
 		SigningTime:      signingTime,
-		HTTPClient:       r.httpClient,
-	})
+		CertChainPurpose: r.certChainPurpose,
+		HTTPClient:       r.ocspHTTPClient,
+	}
+
+	crlOpts := crl.Options{
+		HTTPClient:  r.crlHTTPClient,
+		SigningTime: signingTime,
+	}
+
+	certResults := make([]*result.CertRevocationResult, len(certChain))
+	var wg sync.WaitGroup
+	for i, cert := range certChain[:len(certChain)-1] {
+		switch {
+		case ocsp.SupportOCSP(cert):
+			// do OCSP check for the certificate
+			wg.Add(1)
 
-	// TODO: add CRL support
-	// https://github.com/notaryproject/notation-core-go/issues/125
+			go func(i int, cert *x509.Certificate) {
+				defer wg.Done()
+				ocspResult := ocsp.CertCheckStatus(cert, certChain[i+1], ocspOpts)
+				if ocspResult != nil && ocspResult.Result == result.ResultUnknown && crl.SupportCRL(cert) {
+					// try CRL check if OCSP result is unknown
+					crlResult := crl.CertCheckStatus(r.ctx, cert, certChain[i+1], crlOpts)
+					crlResult.Error = result.OCSPFallbackError{
+						OCSPErr: ocspResult.Error,
+						CRLErr:  crlResult.Error,
+					}
+					certResults[i] = crlResult
+				} else {
+					certResults[i] = ocspResult
+				}
+			}(i, cert)
```

Comment:
Ideally, these detailed steps should be done in their corresponding packages.
Under `revocation` package, the `Validate` function should look like this:
```
func (r *revocation) Validate(certChain []*x509.Certificate, signingTime time.Time) ([]*result.CertRevocationResult, error) {
         switch {
         case support OCSP:
              ocsp.CheckStatus(...)
         case support CRL:
              crl.CheckStatus(...)
         }
}
```